src/similarity.ts

```diff
@@ -13,75 +22,152 @@ import { Headers, PersistedBuffer, TapeRecord } from "./tape";
+  const parsedCompareToQueryParameters = parseQueryParameters(
+    compareTo.request.path,
+  );
+  const differencesQueryParameters = countObjectDifferences(
+    parsedQueryParameters,
+    parsedCompareToQueryParameters,
+    rewriteBeforeDiffRules,
+  );
+
+  // Compare the cleaned headers.
+  const cleanedHeaders = stripExtraneousHeaders(request.headers);
+  const cleanedCompareToHeaders = stripExtraneousHeaders(
     compareTo.request.headers,
   );
-  return (
-    countObjectDifferences(
-      parsedQuery,
-      parsedCompareToQuery,
-      rewriteBeforeDiffRules,
-    ) +
-    countObjectDifferences(headers, compareToHeaders, rewriteBeforeDiffRules) +
-    countBodyDifferences(
-      serialisedRequestBody,
-      serialisedCompareToRequestBody,
-      rewriteBeforeDiffRules,
-    )
+  const differencesHeaders = countObjectDifferences(
+    cleanedHeaders,
+    cleanedCompareToHeaders,
+    rewriteBeforeDiffRules,
   );
+
+  // Compare the bodies.
+  const differencesBody = countBodyDifferences(
+    request,
+    compareTo.request,
+    rewriteBeforeDiffRules,
+  );
+
+  return differencesQueryParameters + differencesHeaders + differencesBody;
 }
 
 /**
- * Returns the numbers of differences between two persisted body buffers.
+ * Returns the numbers of differences between the bodies of two HTTP requests.
  */
 function countBodyDifferences(
-  a: PersistedBuffer,
-  b: PersistedBuffer,
+  request1: HttpRequest,
+  request2: HttpRequest,
   rewriteBeforeDiffRules: RewriteRules,
 ): number {
-  if (a.encoding === "utf8" && b.encoding === "utf8") {
-    try {
-      const requestBodyJson = JSON.parse(a.data || "{}");
-      const recordBodyJson = JSON.parse(b.data || "{}");
-      // Return the number of fields that differ in JSON.
-      return countObjectDifferences(
-        requestBodyJson,
-        recordBodyJson,
-        rewriteBeforeDiffRules,
-      );
-    } catch (e) {
-      return countStringDifferences(a.data, b.data, rewriteBeforeDiffRules);
-    }
+  const contentType1 = parseContentType(getHttpRequestContentType(request1));
+  const contentType2 = parseContentType(getHttpRequestContentType(request1));
+
+  // If the content types are not the same, we cannot compare.
+  if (contentType1.type !== contentType2.type) {
+    return +Infinity;
+  }
+
+  const contentType = contentType1.type;
+  if (contentType === "application/json") {
+    return countBodyDifferencesApplicationJson(
+      request1,
+      contentType1,
+      request2,
+      contentType2,
+      rewriteBeforeDiffRules,
+    );
+  } else if (contentType.startsWith("text/")) {
+    return countBodyDifferencesText(
+      request1,
+      contentType1,
+      request2,
+      contentType2,
+      rewriteBeforeDiffRules,
+    );
+  } else {
+    // No more special cases to consider. Assume binary data for all other content types.
+    return countBodyDifferencesBinary(request1, request2);
+  }
```

Comment:
This is some slightly different logic to before.

Up until now, we just blindly attempted to decode it as JSON. If that succeeded, we did a JS object comparison of the resulting JSON decoded values. If it failed to decode, we did a base64 string comparison.

Now, we inspect the content type of the HTTP requests, and attempt to do JSON decoding if they have `application/json` as their content type, do raw string comparison if they have `text/<anything>` as their content type, and base64 comparison for any other content type.